overlord/devicestate/devicemgr.go

```diff
@@ -2043,3 +2045,73 @@ func (h fdeSetupHandler) Done() error {
 func (h fdeSetupHandler) Error(err error) (bool, error) {
 	return false, nil
 }
+
+var (
+	secbootEnsureRecoveryKey  = secboot.EnsureRecoveryKey
+	secbootRemoveRecoveryKeys = secboot.RemoveRecoveryKeys
+)
+
+// EnsureRecoveryKeys makes sure appropriate recovery keys exist and
+// returns them. Usually a single recovery key is created/used, but
+// older systems might return both a recovery key for ubuntu-data and a
+// reinstall key for ubuntu-save.
+func (m *DeviceManager) EnsureRecoveryKeys() (*client.SystemRecoveryKeysResponse, error) {
+	fdeDir := dirs.SnapFDEDir
+	if m.SystemMode(SysHasModeenv) == "install" {
+		fdeDir = boot.InstallHostFDEDataDir
+	}
+	sysKeys := &client.SystemRecoveryKeysResponse{}
+	// backward compatibility
+	reinstallKeyFile := filepath.Join(fdeDir, "reinstall.key")
+	if osutil.FileExists(reinstallKeyFile) {
+		rkey, err := keys.RecoveryKeyFromFile(filepath.Join(dirs.SnapFDEDir, "recovery.key"))
+		if err != nil {
+			return nil, err
+		}
+		sysKeys.RecoveryKey = rkey.String()
+
+		reinstallKey, err := keys.RecoveryKeyFromFile(reinstallKeyFile)
+		if err != nil {
+			return nil, err
+		}
+		sysKeys.ReinstallKey = reinstallKey.String()
+		return sysKeys, nil
+	}
+	// XXX have a helper somewhere for this? gadget or secboot?
+	if !osutil.FileExists(filepath.Join(fdeDir, "marker")) {
+		return nil, fmt.Errorf("system does not use disk encryption")
+	}
+	dataMountPoints, err := boot.HostUbuntuDataForMode(m.SystemMode(SysHasModeenv))
+	if err != nil {
+		return nil, fmt.Errorf("cannot determine ubuntu-data mount point: %v", err)
+	}
+	mountPoints := []string{dataMountPoints[0], boot.InitramfsUbuntuSaveDir}
+	rkey, err := secbootEnsureRecoveryKey(filepath.Join(fdeDir, "recovery.key"), mountPoints)
+	if err != nil {
+		return nil, err
+	}
+	sysKeys.RecoveryKey = rkey.String()
+	return sysKeys, nil
+}
+
+// RemoveRecoveryKeys removes and disables all recovery keys.
+func (m *DeviceManager) RemoveRecoveryKeys() error {
```

Comment:
I guess that the reason why you don't do that `m.SystemMode(SysHasModeenv) == "install"` check here is that this method is never called in `install` mode?

Reply:
yes, I think it's ok to error if this one is used from install mode (at least for now)